components/raftstore/src/store/fsm/peer.rs

```diff
@@ -1836,25 +1837,90 @@ where
 
     fn on_raft_log_fetched(&mut self, context: GetEntriesContext, res: Box<RaftlogFetchResult>) {
         let low = res.low;
-        // if the peer is not the leader anymore or being destroyed, ignore the result.
-        if !self.fsm.peer.is_leader() || self.fsm.peer.pending_remove {
+        // If the peer is not the leader anymore and it's not about to become leader,
+        // or it is being destroyed, ignore the result.
+        if !self.fsm.peer.is_leader() && self.fsm.peer.pre_become_leader_state.is_none()
+            || self.fsm.peer.pending_remove
+        {
             self.fsm.peer.mut_store().clean_async_fetch_res(low);
             return;
         }
 
         if self.fsm.peer.term() != res.term {
             // term has changed, the result may be not correct.
             self.fsm.peer.mut_store().clean_async_fetch_res(low);
+        } else if !self.fsm.peer.is_leader() {
```

Comment:
I think checking self.fsm.peer.pre_become_leader_state.is_some() is better. 
And also the following block should be wrapped in a func.